### PR TITLE
Make the measurements layer 4D

### PR DESCRIPTION
**Pull request by palvarezc**
For the layer that contains the measurements of the radii and decay lengths, it makes more sense to have the dimensionality of the data itself.

Potentially, this also allows to raise an error if the wrong particle is being used (by consulting the event_number and view_number) introduced in:
- #153 

This is not implemented yet! 

Solves

- #109 
- #165

### CODE REVIEW

**Review by samcunliffe — COMMENTED** [review]
> For the layer that contains the measurements of the radii and decay lengths, it makes more sense to have the dimensionality of the data itself.

But you're not changing the layer setup here? I mean: you've just switched to 4D and haven't used it..?

**Comment by palvarezc** [comment]
> > For the layer that contains the measurements of the radii and decay lengths, it makes more sense to have the dimensionality of the data itself.
> 
> But you're not changing the layer setup here? I mean: you've just switched to 4D and haven't used it..?

Shit, you are right.... I think the tests in `test_calculate` would not catch this.

I think the calculation of lengths is probably fine, but the radius will break. I'll try to update the test...

**Comment by palvarezc** [comment]
Modified the relevant functions to deal with 4D points. In the table and the data we save only the 2D projection (x,y).

**Comment by samcunliffe** [comment]
Retrospectively added 
- https://github.com/palvarezc/cavendish-particle-tracks/issues/165

which is my understanding of the usability feature we're trying to implement here.

**Review by samcunliffe — COMMENTED** [review]
Cool. Tested and works.

I wonder if it's worth trying this with something like the actual data... can you load 1000 events and do this? Presumably `dask` is handling the event loading, but `napari` would be handling 1000 × 3 layers × 3 (or 4) points per event per layer?

**Comment by samcunliffe** [comment]
P.s. is the title of this PR correct? 4D is still the understanding of a layer that is new for each event vs. one layer for all events.

**Comment by samcunliffe** [comment]
> would be handling 1000 × 3 layers × 3 (or 4) points per event per layer?

More like 500 images × 10 points per event × (2  × floating point numbers per point) which is 100s of kbs but should be fine in RAM.

**Review by samcunliffe — APPROVED** [review]

**Comment by samcunliffe** [comment]
Conflicts don't look too bad.

File: src/cavendish_particle_tracks/_main_widget.py [hunk 1]
```diff
@@ -365,7 +384,9 @@ def _on_click_radius(self) -> None:
                 QTableWidgetItem(str(self.data[selected_row].radius_cm)),
             )
 
-            print("Modified particle ", selected_row)
+            napari.utils.notifications.show_info(
+                "Radius added to particle " + str(selected_row)
+            )
```
Comment (palvarezc) [inline src/cavendish_particle_tracks/_main_widget.py:389]: I added this at the end, to show some feedback to the user that the measurement has been recorded. The notifications seem to stay though, as in they do not go away after some time.

Do you think it is annoying?

**Comment by palvarezc** [comment]
> P.s. is the title of this PR correct? 4D is still the understanding of a layer that is new for each event vs. one layer for all events.

(Event, View, x, y)

File: src/cavendish_particle_tracks/_main_widget.py (hunk 1 above)
Comment (palvarezc) [reply src/cavendish_particle_tracks/_main_widget.py:389]: It actually does go away.... it's just a bit slow.

**Pull request merged**